### PR TITLE
[RNMobile] Prevent calling the registerComponent callback multiple times

### LINKED ISSUE #32882: Reusable blocks are displayed as unsupported (development mode only)
## Description

After recent changes (probably related to [the RN `0.64` upgrade](https://github.com/WordPress/gutenberg/pull/29118)), the reusable block is displayed as unsupported even when the block is available. 

## Step-by-step reproduction instructions

1. Use a debug version that connects to the Metro server.
2. Open a post/page that contains a reusable block.
2. Observe that the block is displayed as unsupported.

## Expected behaviour

The reusable block should be available in WP.com sites.

## Actual behaviour

The reusable block is not available.

## Screenshots or screen recording (optional)

Screenshot 2021-06-22 at 12 38 00

## WordPress information
- WordPress version: N/A
- Gutenberg version: N/A
- Are all plugins except Gutenberg deactivated? N/A
- Are you using a default theme (e.g. Twenty Twenty-One)? N/A

## Device information
- Device: iPhone 8 (simulator)
- Operating system: iOS 14.2
- WordPress app version: `779bb171e889ad527cc1c311675119e6596e0127`

### PR DESCRIPTION
**Pull request by fluiddot**
## Description

On development mode, I noticed that the callback executed when registering a component in the App registry is called multiple times. This case happens when the code within the callback triggers any warning or error (this means calling `console.warning` or `console.error`), in [this issue](https://github.com/WordPress/gutenberg/issues/32882#issuecomment-868414379) there's additional helpful information regarding this.

~~This issue has been addressed by adding a static object that contains the components that have been already registered, this way we enforce that the callback's code will be executed only once.~~

**EDIT:** I finally followed a different approach to address this problem, based on the [`AppRegistry` documentation](https://reactnative.dev/docs/appregistry), I created a class component that holds the initialization logic.

## How has this been tested?

1. Add the following code `debugger;` before [this line](https://github.com/WordPress/gutenberg/blob/dc27a49391947e2430010956a66afab4d95cc1d7/packages/element/src/react-platform.native.js#L25)
2. Run the app connected with Metro and debug it with Chrome
3. Open a post/page
4. Observe that the debugger gets paused due to the code above.
5. Resume the script execution Screenshot 2021-06-25 at 13 35 20
6. Observe that the debugger is not paused again.

## Verify that reusable blocks are rendered

The original issue was spotted due to the reusable blocks being rendered as unsupported blocks so it would be nice to verify that the issue is no longer happening.

1. Open a post/page that contains reusable blocks (if required add them via the web version of the editor)
2. Observe that the reusable blocks are rendered properly

## Screenshots 
N/A

## Types of changes

Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by hypest** [comment]
> This issue has been addressed by adding a static object that contains the components that have been already registered, this way we enforce that the callback's code will be executed only once.

👋 @fluiddot ! Gave a first look at the solution here and even though I understand how it works, I don't feel particularly confident about fighting against the current lifecycle of the app and components by keeping a static variable. I feel that this is a indication that the lifecycle is not clean enough (on one hand) and that adding a static variable to combat it will only move the problem elsewhere. I understand that this feeling is not super specific or concrete though.

Since the issue here only happens in the development build, I wonder if we should just spend some more time finding a different solution, one that would probably leave the component lifecycle in a better state? Let me know what you think.

**Comment by fluiddot** [comment]
> > This issue has been addressed by adding a static object that contains the components that have been already registered, this way we enforce that the callback's code will be executed only once.
> 
> 👋 @fluiddot ! Gave a first look at the solution here and even though I understand how it works, I don't feel particularly confident about fighting against the current lifecycle of the app and components by keeping a static variable. I feel that this is a indication that the lifecycle is not clean enough (on one hand) and that adding a static variable to combat it will only move the problem elsewhere. UI understand that this feeling is not super specific or concrete though.
> 
> Since the issue here only happens in the development build, I wonder if we should just spend some more time finding a different solution, one that would probably leave the component lifecycle in a better state? Let me know what you think.

My initial approach when I tackled this issue was to figure out why we have some differences between the production and development modes, like the reusable blocks being unsupported. After the investigation, I realized that was related to the App component's lifecycle and logging errors/warnings, so I wanted to propose this PR as a workaround but I agree that is more like a hack than an actual solution. I have some ideas that I'd like to explore so I'll set the PR to draft and propose a new solution.

Thanks for reviewing it 🙇 !

File: packages/element/src/react-platform.native.js [hunk 1]
```diff
@@ -8,27 +8,39 @@ import { omit } from 'lodash';
...
+			this.filteredProps = applyFilters(
+				'native.block_editor_props',
+				parentProps
+			);
+		}
 
-		return cloneElement( element, filteredProps );
-	} );
+		componentDidMount() {
+			doAction( 'native.render', this.filteredProps );
```
Comment (fluiddot) [inline packages/element/src/react-platform.native.js:33]: I understand that the action `native.render` should happen after the component is rendered, so now this will be executed after the component is mounted, this means after the first render.

File: packages/element/src/react-platform.native.js [hunk 2]
```diff
@@ -8,27 +8,39 @@ import { omit } from 'lodash';
...
+			super( ...arguments );
 
-		doAction( 'native.pre-render', parentProps );
+			const parentProps = omit( this.props || {}, [ 'rootTag' ] );
 
-		const filteredProps = applyFilters(
-			'native.block_editor_props',
-			parentProps
-		);
+			doAction( 'native.pre-render', parentProps );
```
Comment (fluiddot) [inline packages/element/src/react-platform.native.js:24]: The `native.pre-render` action will be executed right before the component is mounted, this means before the first render.

File: packages/element/src/react-platform.native.js [hunk 3]
```diff
@@ -8,27 +8,39 @@ import { omit } from 'lodash';
...
 /**
  * Internal dependencies
  */
 import { cloneElement } from './react';
 
-const render = ( element, id ) =>
-	AppRegistry.registerComponent( id, () => ( propsFromParent ) => {
-		const parentProps = omit( propsFromParent || {}, [ 'rootTag' ] );
+const render = ( element, id ) => {
+	class App extends Component {
```
Comment (fluiddot) [inline packages/element/src/react-platform.native.js:18]: The `App` class component is defined within the `render` function because it uses the `element` argument to render the component.

**Comment by fluiddot** [comment]
Since this part is quite critical I wanted to have more 👀 to review this, @geriux @dcalhoun I'd appreciate it if you could take a look at this refactor of the initialization logic of the editor in RN, thanks 🙇 !

File: packages/element/src/react-platform.native.js (hunk 2 above)
Comment (dcalhoun) [reply packages/element/src/react-platform.native.js:24]: I do not follow exactly how the "ceremony" of a class addresses the original issue of multiple calls to `native.pre-render`. Is `render`/`AppRegistry.registerComponent` still called multiple times? Is the class instance retained between the calls?

**Review by dcalhoun — COMMENTED** [review]
👍🏻 Verified the the `debugger` was only hit once and Reusable Blocks functioned on both an iPhone SE and Samsung Galaxy S20.

File: packages/element/src/react-platform.native.js (hunk 2 above)
Comment (fluiddot) [reply packages/element/src/react-platform.native.js:24]: As far as I checked, the multiple calls are caused when returning a function instead of a component when registering the component in AppRegistry. For example, if you replace line 41 with the following code:

```
	AppRegistry.registerComponent( id, () => ( props ) => {
		debugger;
		return <App { ...props } />;
	} );
```

The debugger will pause for every warning and error that happens during the initialization, I haven't located yet the exact part from the React Native code that causes this. My impression is that when a warning/error is handled and tries to get the component stack, for some reason is calling the function returned from the callback passed in the `AppRegistry.registerComponent` call.

I haven't checked with a functional component, but in this case, since we need to execute the pre-render action on the constructor, I haven't found a reliable way to do it with a functional component so I used a classic component.

Comment (dcalhoun) [reply packages/element/src/react-platform.native.js:24]: > The debugger will pause for every warning and error that happens during the initialization, I haven't located yet the exact part from the React Native code that causes this. My impression is that when a warning/error is handled and tries to get the component stack, for some reason is calling the function returned from the callback passed in the `AppRegistry.registerComponent` call.

Ah, I see. From your statement and debugging myself a little further my perception is that the component registered with `AppRegistry` is merely re-rendering when these warnings/errors occur. I presume that is expected, but I have no explicit source to back up that presumption. However, I now understand the `doAction` calls to be side effects and shouldn't be invoked on every render. 

> I haven't checked with a functional component, but in this case, since we need to execute the pre-render action on the constructor, I haven't found a reliable way to do it with a functional component so I used a classic component.

From a brief test, wrapping the side effects with `useEffect` does appear to have the same outcome, i.e. preventing multiple runs, but there is not explicit reason to avoid a class. So, leveraging the class approach is OK with me. 

<details><summary>Functional App</summary>

```jsx
/**
 * External dependencies
 */
import { AppRegistry } from 'react-native';
import { omit } from 'lodash';

/**
 * WordPress dependencies
 */
import { applyFilters, doAction } from '@wordpress/hooks';
import { useEffect } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { cloneElement } from './react';

const render = ( element, id ) =>
	AppRegistry.registerComponent( id, () => ( propsFromParent ) => {
		const parentProps = omit( propsFromParent || {}, [ 'rootTag' ] );

		useEffect( () => {
			doAction( 'native.pre-render', parentProps );
		}, [] );

		const filteredProps = applyFilters(
			'native.block_editor_props',
			parentProps
		);

		useEffect( () => {
			doAction( 'native.render', filteredProps );
		}, [] );

		return cloneElement( element, filteredProps );
	} );

/**
 * Render a given element on Native.
 * This actually returns a componentProvider that can be registered with `AppRegistry.registerComponent`
 *
 * @param {WPElement} element Element to render.
 */
export { render };
```

</details>

In regards to "we need to execute the pre-render action on the constructor," is this to imply that `doAction( 'native.render', filteredProps );` (or other code) is dependent upon `doAction( 'native.pre-render', parentProps );`? Or was your statement merely your interpretation based upon the existing name "pre-render"?

Comment (fluiddot) [reply packages/element/src/react-platform.native.js:24]: > > The debugger will pause for every warning and error that happens during the initialization, I haven't located yet the exact part from the React Native code that causes this. My impression is that when a warning/error is handled and tries to get the component stack, for some reason is calling the function returned from the callback passed in the `AppRegistry.registerComponent` call.
> 
> Ah, I see. From your statement and debugging myself a little further my perception is that the component registered with `AppRegistry` is merely re-rendering when these warnings/errors occur. I presume that is expected, but I have no explicit source to back up that presumption. However, I now understand the `doAction` calls to be side effects and shouldn't be invoked on every render.

Exactly, the `doAction` calls should be treated as side effects and only executed once.

Regarding why the function is being called, it's a bit hard to follow the trace but it's related to how the `react-devtools` package builds the component stacktrace when a warning/error is triggered, here are some screenshots to clarify it:

**1. Log warning is triggered**
log-warning

**2. Getting the component stack (`react-dev-tools`)**

The important part here is the `current` variable that holds the current Fiber node and is passed down the stack trace.
fibernode

Here you can see that the `type` property of the current Fiber node is the App function:
fibernode-type

**3. Describe function component frame**

The previous `current` value is passed as the `workInProgress` argument in the function shown in the screenshot, note that the first argument of the call to `describeFunctionComponentFrame` is the `type` value (the App function) that was mentioned previosly.
describe-fiber

**4. Describe native component frame - App function call**
fn-call
 
> > I haven't checked with a functional component, but in this case, since we need to execute the pre-render action on the constructor, I haven't found a reliable way to do it with a functional component so I used a classic component.
> 
> From a brief test, wrapping the side effects with `useEffect` does appear to have the same outcome, i.e. preventing multiple runs, but there is not explicit reason to avoid a class. So, leveraging the class approach is OK with me.

Thanks for sharing the functional component approach 🙇 ! I agree that wrapping the `doAction` calls with `useEffect` is enough to prevent calling them multiple times, but I decided to go with the class approach because I thought that the `native.pre-render` call should happen before the component is rendered and `useEffect` is triggered after the first render. Due to this reason is why I added the `native.pre-render` action in the class contructor, I found some workarounds to run code on a functional component before the first render but looked like hacks.

> In regards to "we need to execute the pre-render action on the constructor," is this to imply that `doAction( 'native.render', filteredProps );` (or other code) is dependent upon `doAction( 'native.pre-render', parentProps );`? Or was your statement merely your interpretation based upon the existing name "pre-render"?

As far as I checked in [the file where the actions are defined](https://github.com/wordpress-mobile/gutenberg-mobile/blob/864c1a9c2687cb842060aa06be7cab1f957904da/src/index.js#L18-L50), I don't feel there are any dependency between the actions but I infer from the action's name `native.pre-render` that this should be called before the editor is rendered 🤔 .

Comment (dcalhoun) [reply packages/element/src/react-platform.native.js:24]: > Regarding why the function is being called, it's a bit hard to follow the trace but it's related to how the `react-devtools` package builds the component stacktrace when a warning/error is triggered [...]

Yeah, I did notice — if I remembering correctly — the original multiple side effect calls issue only occurs if Debug is enabled from the developer menu. 

> As far as I checked in [the file where the actions are defined](https://github.com/wordpress-mobile/gutenberg-mobile/blob/864c1a9c2687cb842060aa06be7cab1f957904da/src/index.js#L18-L50), I don't feel there are any dependency between the actions but I infer from the action's name `native.pre-render` that this should be called before the editor is rendered 🤔 .

Makes sense. I have no data to suggest otherwise at this point. The class constructor approach does ensure it invokes before initial render.

**Review by dcalhoun — APPROVED** [review]
LGTM. 👍🏻 My earlier https://github.com/WordPress/gutenberg/pull/32985#pullrequestreview-704533708 verified the app functioned as expected. Thank you for sorting this out. 🙇🏻

File: packages/element/src/react-platform.native.js (hunk 2 above)
Comment (fluiddot) [reply packages/element/src/react-platform.native.js:24]: > Yeah, I did notice — if I remembering correctly — the original multiple side effect calls issue only occurs if Debug is enabled from the developer menu.

I think it also happens when the Debug is disabled because the reusable blocks are still not showing up in that case. I infer it's being called multiple times because the reusable block is being unregistered when this code gets executed a second time, surprisingly the `capabilities` object is empty and since the reusable block has been already registered, the condition is true. 

https://github.com/WordPress/gutenberg/blob/db76f671b07a0fab33de969b3235ca181c5d3a42/packages/react-native-editor/src/index.js#L60-L66

> Makes sense. I have no data to suggest otherwise at this point. The class constructor approach does ensure it invokes before initial render.

👍
